src/caching.ts

```diff
@@ -0,0 +1,99 @@
+import { MemoryCache, MemoryConfig, MemoryStore, memoryStore } from './stores';
+
+export interface Config {
+  ttl?: Ttl;
+  isCacheable?: (val: unknown) => boolean;
+}
+
+export type Ttl = number;
+
+export interface Store {
+  get<T>(key: string): Promise<T | undefined>;
+  set<T>(key: string, data: T, ttl?: Ttl): Promise<void>;
+  del(key: string): Promise<void>;
+  del(key: string[]): Promise<void>;
+  del(key: string[] | string): Promise<void>;
+  reset(): Promise<void>;
+  mset(args: [string, unknown][], ttl?: Ttl): Promise<void>;
+  mget(...args: string[]): Promise<unknown[]>;
+  keys(): Promise<string[]>;
+  keys(pattern: string): Promise<string[]>;
+  ttl(key: string): Promise<number>;
+}
+
+export type StoreConfig = Config;
+
+export type FactoryConfig<T> = T & Config;
+type FactoryStore<S extends Store = Store, T extends object = never> = (
+  config?: FactoryConfig<T>,
+) => S | Promise<S>;
+
+type Stores<S extends Store, T extends object> =
+  | 'memory'
+  | Store
+  | FactoryStore<S, T>;
+type CachingConfig<T> = MemoryConfig | StoreConfig | FactoryConfig<T>;
+
+export interface Cache<S extends Store = Store> {
+  set: S['set'];
+  get: S['get'];
+  del: S['del'];
+  reset: S['reset'];
+  wrap<T>(key: string, fn: () => Promise<T>): Promise<T>;
+  store: S;
+}
+
+export async function caching(
+  store: 'memory',
+  args?: MemoryConfig,
+): Promise<MemoryCache>;
+export async function caching<S extends Store = Store>(
+  store: S,
+): Promise<Cache<S>>;
+export async function caching<
+  S extends Store = Store,
+  T extends object = never,
+>(store: FactoryStore<S, T>, args?: FactoryConfig<T>): Promise<Cache<S>>;
+
+/**
+ * Generic caching interface that wraps any caching library with a compatible interface.
+ */
+export async function caching<
+  S extends Store = Store,
+  T extends object = never,
+>(fac: Stores<S, T>, args?: CachingConfig<T>): Promise<Cache<S | MemoryStore>> {
```

Comment:
I'd personally not use `fac`, but instead use the full word `factory`, etc.

Reply:
I'd be hesitant to make the `caching` function (which is pretty much a constructor) be asynchronous.  Some users may want/need to instantiate their cache instances synchronously.  Those cases are probably just due to legacy code, however, so maybe that's not an issue.

Reply:
> I'd personally not use `fac`, but instead use the full word `factory`, etc.

Ok, I will change this variable names

Reply:
Looks like the caching function being `async` is causing some issues (https://github.com/node-cache-manager/node-cache-manager/issues/210#issuecomment-1263920136).  Maybe the constructor should be synchronous and a separate method like `initialize()` or something can be async if needed.